velox/row/UnsafeRowFast.cpp

```diff
@@ -0,0 +1,369 @@
+
+int32_t alignBits(int32_t numBits) {
+  return bits::nwords(numBits) * 8;
+}
+
+int32_t alignBytes(int32_t numBytes) {
+  return bits::roundUp(numBytes, 8);
+}
+} // namespace
+
+// static
+std::optional<int32_t> UnsafeRowFast::fixedRowSize(const RowTypePtr& rowType) {
+  for (const auto& child : rowType->children()) {
+    if (!child->isFixedWidth()) {
+      return std::nullopt;
+    }
+  }
+
+  const size_t numFields = rowType->size();
+  size_t nullLength = alignBits(numFields);
+
+  return nullLength + numFields * kFieldWidth;
+}
+
+UnsafeRowFast::UnsafeRowFast(const RowVectorPtr& vector)
+    : typeKind_{vector->typeKind()}, decoded_{*vector} {
+  initialize(vector->type());
+}
+
+UnsafeRowFast::UnsafeRowFast(const VectorPtr& vector)
+    : typeKind_{vector->typeKind()}, decoded_{*vector} {
+  initialize(vector->type());
+}
+
+void UnsafeRowFast::initialize(const TypePtr& type) {
+  auto base = decoded_.base();
+  switch (typeKind_) {
+    case TypeKind::ARRAY: {
+      auto arrayBase = base->as<ArrayVector>();
+      children_.push_back(UnsafeRowFast(arrayBase->elements()));
+      fixedWidth_.push_back(arrayBase->elements()->type()->isFixedWidth());
+      break;
+    }
+    case TypeKind::MAP: {
+      auto mapBase = base->as<MapVector>();
+      children_.push_back(UnsafeRowFast(mapBase->mapKeys()));
+      children_.push_back(UnsafeRowFast(mapBase->mapValues()));
+      fixedWidth_.push_back(mapBase->mapKeys()->type()->isFixedWidth());
+      fixedWidth_.push_back(mapBase->mapValues()->type()->isFixedWidth());
+      break;
+    }
+    case TypeKind::ROW: {
+      auto rowBase = base->as<RowVector>();
+      for (const auto& child : rowBase->children()) {
+        children_.push_back(UnsafeRowFast(child));
+        fixedWidth_.push_back(child->type()->isFixedWidth());
+      }
+
+      rowNullBytes_ = alignBits(type->size());
+      break;
+    }
+    case TypeKind::BOOLEAN:
+      valueBytes_ = 1;
+      break;
+    case TypeKind::TINYINT:
+    case TypeKind::SMALLINT:
+    case TypeKind::INTEGER:
+    case TypeKind::BIGINT:
+    case TypeKind::REAL:
+    case TypeKind::DOUBLE:
+    case TypeKind::DATE:
+      valueBytes_ = type->cppSizeInBytes();
+      supportsBulkCopy_ = decoded_.isIdentityMapping();
+      break;
+    case TypeKind::TIMESTAMP:
+      valueBytes_ = sizeof(int64_t);
+      break;
+    default:
+        // Nothing to do.
+        ;
+  }
+}
+
+int32_t UnsafeRowFast::rowSize(vector_size_t index) {
+  return rowRowSize(index);
+}
+
+int32_t UnsafeRowFast::variableWidthRowSize(vector_size_t index) {
+  switch (typeKind_) {
+    case TypeKind::VARCHAR:
+    case TypeKind::VARBINARY: {
+      auto value = decoded_.valueAt<StringView>(index);
+      return alignBytes(value.size());
+    }
+    case TypeKind::ARRAY:
+      return arrayRowSize(index);
+    case TypeKind::MAP:
+      return mapRowSize(index);
+    case TypeKind::ROW:
+      return rowRowSize(index);
+    default:
+      VELOX_UNREACHABLE();
```

Comment:
Can we also log the unexpected kind_ and row index in error message for debugging?